### PR TITLE
Generate XSLT StyleSheet - DataMapper

### PR DESCRIPTION
**Pull request by DMadhuranga**
## Purpose

Creating an additional artifact when creating a datamapper diagram and generate an XSLT StyleSheet when drawing a datamapper diagram. A new property (XSLT StyleSheet) will be added to the datamapper.
This new artifact will also be packed in to the carbon app when exporting.

Related Issue : wso2/product-ei#2745

## Goals

Increase the performance of XML to XML message transformaion in datamapper mediator.

Related Issue : wso2/product-ei#2745

## Related PRs

Carbon Mediation : https://github.com/wso2/carbon-mediation/pull/1096

All the datamapper mediators created up to version 6.4.0 will be supported.

## Documentation

Documentation Issue : https://github.com/wso2/product-ei/issues/2746

### Problem

Currently, XML to XML mapping is done through a Javascript engine.

1. First, the input XML file is converted to a JSON file.
2. Then, an output JSON file is created using the generated input JSON file. (JSON-to-JSON mapping. This part is performed by the JavaScript engine)
3. Finally, the output JSON file is converted to a XML file. (Output XML file)

The main issue is the poor performance due to multiple transformations across formats and performance of the JavaScript engine. Therefore, it was proposed to introduce direct XML-to-XML mapping without undergoing these transformations and processing through the JavaScript engine.

### Design

Implement a library to auto-generate a XSLT file from the diagrams of the developer studio. Then transform input XML to output XML using an XSLT transformation engine. An XSLT transformation engine already exists in Java and the EI.

Many companies are already using the above method         
There are libraries available to convert XML to other formats such as JSON using XSLT.
Since XSLT is a standard styling language for XML transformation, generating a XSLT file could increase interoperability of the datamapper.

mydiagram12 2

#### Performance Test Prior To Implementation

I conducted a performance test comparing performance of current data mapper mediator (XML-to-XML transformation) and the  XSLT mediator performing the same mapping. I measure the response time of each request of the two transformation methods while increasing the number of concurrent threads that send requests. Test result was as following.

plot 5

In average, XSLT transformation was roughly twice as fast compared with data mapper.

### Implementation

Previously, when an user draw a datamapper diagram in the developer studio, five main artifacts get generated. They are “InputSchema.js” file, “OutputSchema.js” file, “.dmc” file, “.datamapper” file (Datamapper schema) and “.datamapper_diagram” file.

“InputSchema.js” file represents the model of the input message
“OutputSchema.js” file represents the model of the output message
“.dmc” file contains the JavaScript function required to transform the input to output in the server.
Datamapper schema contains the details on mapping configuration
“.datamapper_diagram” file contains the details of the graphical objects in the diagram

A new artifact will be generated by the developer studio in addition to the current artifacts in order to support the XSLT based transformation. The new artifact is an XSLT stylesheet. It is generated by the “XSLT generator library” 

#### XSLT generator library

When a user draws a diagram to configure the mapping between input XML and output XML, the developer studio saves the details in the diagram in a XML file (Datamapper schema). The library reads the Datamapper schema and generates the XSLT StyleSheet. In the server input, XML messages are transformed to the desired output XML message by passing the XSLT Stylesheet and the input XML to an XSLT Engine (Saxon engine). Following diagram explains the main operations of the library.

library

##### Datamapper Schema

When an user draw a diagram to configure mapping between input XML and output XML, the developer studio saves the details in the diagram in a XML file (Datamapper schema). Thus, this library takes the Datamapper schema and generate a XSLT style-sheet.  

Datamapper diagram contains two tree structures and a list. The first tree structure is the structure representing the structure of input message. The second tree structure is the structure representing the structure of output message. And the end there is a list of nodes each node representing an operator of the datamapper diagram. There are links to output node from the input nodes and the operator nodes representing the mapping.

Consider the following simple mapping scenario.

diagram

This mapping configuration is saved in the datamapper schema as follows. Note that this is a simplified version of the datamapper schema.

```xml
<input>
   <node name="student" type="object">
       <node name="name" type="string"></node>
       <node name="age" type="number"></node>
       <node name="school" type="object">
           <node name="school_name" type="string"></node>
           <node name="address" type="string"></node>
       </node>
   </node>
</input>

<output>
   <node name="user" type="object">
      <node name="username" type="string"><incomingLink>input/node.0/node.0</incomingLink></node>
      <node name="age" type="number"><incomingLink>input/node.0/node.1</incomingLink></node>
      <node name="school" type="object">
         <node name="school_name"type="string"> 
            <incomingLink>input/node.0/node.2/node.0</incomingLink>
         </node>
         <node name="address" type="string">
            <incomingLink>input/node.0/node.2/node.1</incomingLink>
         </node>
      </node>
   </node>
</output>
```

In the datamapper schema, each node represents an entity in the input or output schema. In addition to a name property, each node contains a type property indicating what will be inside the node. String, Number and Object are few types of nodes. Object node contains other nodes, but String and Number type nodes contain only values. There is a link from each node of the output to an input node or an operator node. Since there are no operator nodes in this example, only links to input nodes are visible.

##### Initialization of generation

The XSLT generator library also have two tree structures and an array data structure. An input tree representing the input tree in the datamapper schema, an output tree representing the output tree in the datamapper schema, and an array of operators representing operators array in the datamapper schema. Prior to generating the XSLT Stylesheet, the library scan the Datamapper schema and generate input tree, output tree and the list of operators.

There are many operators in the datamapper. And the XSLT transformation supports all the operators except the “Custom Function” operator. After generating the data structures, it checks the array of operators for any “Custom Function” operator. If there exists a “Custom Function” operator, it is not possible to generate an XSLT StyleSheet. Therefore, aborts the generation.

##### Traverse through output tree and generate XSLT Stylesheet

After generating all the required data structures (Input tree, Output tree and the list of operators), the library traverse through the output tree and generate XSLT Stylesheet. In this process, it traverses through input tree and list of operators when required to get details related to mapping. Generated XSLT file is as follows.

```xml
<user>
   <username><xsl:value-of select="student/name"/></username>
   <age><xsl:value-of select="number( student/age )"/></age>
   <school>
       <school_name><xsl:value-of select="student/school/school_name"/></school_name>
       <address><xsl:value-of select="student/school/address"/></address>
   </school>
</user>
```

Note that this is a simplified representation of the XSLT file. When traversing through the output tree, the root node (user node) will be processed first. Since the type of the user node is an object, simply a “<user></user>” tag gets added and processing start on its child nodes. Everything gets added to the XSLT stylesheet afterward will be added into the user tag. Processing the first child node (“username”) of the user node, adding “xsl:value-of” tag and adding XPath to the mapping input node.

### CODE REVIEW

**Comment by a5anka** [comment]
@prabushi Can you help to review this PR?

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTStyleSheetGenerationHandler.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+     * @throws ParserConfigurationException 
+     * @throws IOException 
+     * @throws SAXException 
+     */
+    public String generate(InputStream schemaInPutSource) throws TransformerException, SAXException, IOException, ParserConfigurationException {
+        DataMapperSchemaProcessor inputXML = new DataMapperSchemaProcessor(schemaInPutSource);
+        XSLTStyleSheetWriter outputXML = new XSLTStyleSheetWriter();
+        XSLTGenerator xsltGenerator = new XSLTGenerator();
+        boolean xsltApplicable = xsltGenerator.initializeGeneration(inputXML, outputXML);
+        if (xsltApplicable) {
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTStyleSheetGenerationHandler.java]: Shall we define why we are skipping this?

Comment (abeykoon) [reply plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTStyleSheetGenerationHandler.java]: How are we informing the user there is some incompatibility at this point?

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/xmltree/OutPutNode.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+     * This method extracts all the properties and build child nodes
+     * 
+     * @param node document node that represents the output node
+     */
+    private void populate(Node node) {
+        for (int i = 0; i < node.getChildNodes().getLength(); i++) {
+            Node childNode = node.getChildNodes().item(i);
+            if (childNode.getAttributes() != null) {
+                switch (childNode.getNodeName()) {
+                case PROPERTIES_LOWER_CASE:
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/xmltree/OutPutNode.java]: What operators are defined as "PROPERTIES_LOWER_CASE". As discussed on the review better use a meaningful name.

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/config/XSLTGeneratorConstants.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    public static final String DEFAULT_VALUE = "defaultValue";
+    public static final String DEFAULT_NAME = "defaultName";
+    public static final String CUSTOM_FUNCTION = "CUSTOM_FUNCTION";
+    public static final String FUNCTION_DEFINITION = "functionDefinition";
+    public static final String XSL_NAMESPACE_URI = "http://www.w3.org/1999/XSL/Transform";
+    public static final String XS_NAMESPACE_URI = "http://www.w3.org/2001/XMLSchema";
+    public static final String XSLT_VERSION = "2.0";
+    public static final String XSLT_FUNCTION_DECLARE_URI = "http://ownfunction.org";
+    public static final String RUN_TIME_PROPERTIES = "xmlns:runTimeProperties";
+    public static final String NOT_XSLT_COMPATIBLE = "xmlns:notXSLTCompatible";
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/config/XSLTGeneratorConstants.java]: Shall we rename this to "XSLT_COMPATIBLE" for readability and change logics accordingly?

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/config/XSLTGeneratorConstants.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    private XSLTGeneratorConstants() {
+    }
+
+    // constants
+    public static final String STRING_TYPE = "string";
+    public static final String OBJECT_TYPE = "object";
+    public static final String ARRAY_TYPE = "array";
+    public static final String NUMBER_TYPE = "number";
+    public static final String BOOLEAN_TYPE = "boolean";
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/config/XSLTGeneratorConstants.java]: Shall we define types (relation to nodes in .datamapper file?)
i.e array type is a node having child nodes.

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/part/DataMapperDiagramEditor.java [hunk 5]
```diff
@@ -496,6 +502,38 @@ private boolean updateAssociatedConfigFile(IProgressMonitor monitor) {
 		}
 		return true;
 	}
+	
+	private void updateAssociatedXsltFile(IProgressMonitor monitor) {
+		IEditorInput editorInput = this.getEditorInput();
+        if (editorInput instanceof IFileEditorInput) {
+            IFile diagramFile = ((FileEditorInput) editorInput).getFile();
+            String diagramFilePath = diagramFile.getFullPath().toString();
+            String schemaFilePath = diagramFilePath.replaceAll(".datamapper_diagram$", ".datamapper"); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (prabushi) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/part/DataMapperDiagramEditor.java]: Remove comments (i.e. //@NON-NLS-1$ etc)

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/part/DataMapperDiagramEditor.java [hunk 6]
```diff
@@ -496,6 +502,38 @@ private boolean updateAssociatedConfigFile(IProgressMonitor monitor) {
...
+            String xsltStyleSheetFilePath = diagramFilePath.replaceAll(".datamapper_diagram$", "_xsltStyleSheet.xml");
+            IFile schemaFile = diagramFile.getWorkspace().getRoot().getFile(new Path(schemaFilePath));
+            IFile xsltStyleSheetFile = diagramFile.getWorkspace().getRoot().getFile(new Path(xsltStyleSheetFilePath));
+            InputStream schemaInPutStream = null;
+            if (xsltStyleSheetFile.exists() && schemaFile.exists()) {
+                try {
```
Comment (prabushi) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/part/DataMapperDiagramEditor.java]: We can use try with resource and remove the finally block. https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTGenerator.java [hunk 7]
```diff
@@ -0,0 +1,1017 @@
...
+    private void createOperatorNodes(DataMapperSchemaProcessor inputXMLFile) {
+        operatorNodes = new ArrayList<>();
+        globalVariables = new ArrayList<>();
+        NodeList operators = inputXMLFile.getDocument().getElementsByTagName("operators");
+        for (int i = 0; i < operators.getLength(); i++) {
+            operatorNodes.add(new OperatorNode(operators.item(i)));
+        }
+    }
+
+    private boolean setPropertyOperators(XSLTStyleSheetWriter outputXMLFile) {
```
Comment (prabushi) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTGenerator.java]: Add a method description explaining the cases addressed here.

File: plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTStyleSheetGenerationHandler.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import org.wso2.developerstudio.datamapper.diagram.xslt.config.DataMapperSchemaProcessor;
+import org.wso2.developerstudio.datamapper.diagram.xslt.config.XSLTStyleSheetWriter;
+import org.xml.sax.SAXException;
+
+/**
+ * This class handle generation of the xslt stylesheet.
+ */
+public class XSLTStyleSheetGenerationHandler {
+
+    public static XSLTStyleSheetGenerationHandler instance;
```
Comment (prabushi) [inline plugins/org.wso2.developerstudio.visualdatamapper.diagram/src/org/wso2/developerstudio/datamapper/diagram/xslt/XSLTStyleSheetGenerationHandler.java]: Make a private instance.

**Comment by a5anka** [comment]
@DMadhuranga @prabushi Are we done with the review?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.edit/src/org/wso2/developerstudio/eclipse/gmf/esb/provider/DataMapperMediatorItemProvider.java [hunk 9]
```diff
@@ -308,42 +294,45 @@ public Object getImage(Object object) {
...
+        case EsbPackage.DATA_MAPPER_MEDIATOR__OUTPUT_TYPE:
+            fireNotifyChanged(new ViewerNotification(notification, notification.getNotifier(), false, true));
+            return;
+        case EsbPackage.DATA_MAPPER_MEDIATOR__INPUT_CONNECTOR:
+        case EsbPackage.DATA_MAPPER_MEDIATOR__OUTPUT_CONNECTOR:
+        case EsbPackage.DATA_MAPPER_MEDIATOR__CONFIGURATION:
+        case EsbPackage.DATA_MAPPER_MEDIATOR__INPUT_SCHEMA:
+        case EsbPackage.DATA_MAPPER_MEDIATOR__OUTPUT_SCHEMA:
+        case EsbPackage.DATA_MAPPER_MEDIATOR__XSLT_STYLE_SHEET:
+            fireNotifyChanged(new ViewerNotification(notification, notification.getNotifier(), true, false));
```
Comment (rosensilva) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.edit/src/org/wso2/developerstudio/eclipse/gmf/esb/provider/DataMapperMediatorItemProvider.java]: Are we intentionally going to change the notification firing from case EsbPackage.DATA_MAPPER_MEDIATOR__OUTPUT_SCHEMA to case EsbPackage.DATA_MAPPER_MEDIATOR__XSLT_STYLE_SHEET ?

Comment (DMadhuranga) [reply plugins/org.wso2.developerstudio.eclipse.gmf.esb.edit/src/org/wso2/developerstudio/eclipse/gmf/esb/provider/DataMapperMediatorItemProvider.java]: This is a generated code block.

**Comment by pavanandhukuri** [comment]
Has this change been released? I don't see anything in the documentation related to this feature.

**Pull request merged**